portability-gateway/src/main/java/org/dataportabilityproject/gateway/reference/DataTransferHandler.java

```diff
@@ -0,0 +1,177 @@
+      ApiSettings apiSettings,
+      JobStore store,
+      TypeManager typeManager,
+      SymmetricKeyGenerator symmetricKeyGenerator) {
+    this.createJobAction = createJobAction;
+    this.registry = registry;
+    this.apiSettings = apiSettings;
+    this.store = store;
+    this.symmetricKeyGenerator = symmetricKeyGenerator;
+    this.objectMapper = typeManager.getMapper();
+  }
+
+  /**
+   * Services the {@link CreateJobAction} via the {@link HttpExchange}.
+   */
+  @Override
+  public void handle(HttpExchange exchange) throws IOException {
+    Preconditions.checkArgument(
+        ReferenceApiUtils.validateRequest(exchange, HttpMethods.POST, PATH),
+        PATH + " only supports POST.");
+    logger.debug("received request: {}", exchange.getRequestURI());
+    DataTransferRequest request = objectMapper
+        .readValue(exchange.getRequestBody(), DataTransferRequest.class);
+    CreateJobActionRequest actionRequest = new CreateJobActionRequest(request.getSource(),
+        request.getDestination(), request.getTransferDataType());
+    CreateJobActionResponse actionResponse = createJobAction.handle(actionRequest);
+
+    DataTransferResponse dataTransferResponse;
+    if (actionResponse.getErrorMsg() != null) {
+      logger.warn("Error during action: {}", actionResponse.getErrorMsg());
+      handleError(exchange, request);
+      return;
+    }
+
+    // Set new cookie
+    String encodedJobId = ReferenceApiUtils.encodeId(actionResponse.getId());
+    HttpCookie cookie = new HttpCookie(JsonKeys.ID_COOKIE_KEY, encodedJobId);
+    exchange.getResponseHeaders()
+        .add(HttpHeaders.SET_COOKIE, cookie.toString() + ReferenceApiUtils.COOKIE_ATTRIBUTES);
+
+    // Initial auth flow url
+    AuthDataGenerator generator = registry.getAuthDataGenerator(request.getSource(), request.getTransferDataType(),
+        AuthMode.EXPORT);
+    Preconditions.checkNotNull(generator, "Generator not found for type: %s, service: %s",
+        request.getTransferDataType(), request.getSource());
+
+    AuthFlowConfiguration authFlowConfiguration = generator
+        .generateConfiguration(apiSettings.getBaseUrl(), encodedJobId);
+    Preconditions.checkNotNull(authFlowConfiguration,
+        "AuthFlowConfiguration not found for type: %s, service: %s",
+        request.getTransferDataType(), request.getSource());
+
+    PortabilityJob job = store.findJob(actionResponse.getId());
+
+    // If present, store initial auth data for export services, e.g. used for oauth1
+    if (authFlowConfiguration.getInitialAuthData() != null) {
+
+      // Retrieve and parse the session key from the job
+      String sessionKey = job.jobAuthorization().encryptedSessionKey();
+      SecretKey key = symmetricKeyGenerator.parse(BaseEncoding.base64Url().decode(sessionKey));
+
+      // Ensure intial auth data for export has not already been set
+      Preconditions.checkState(Strings.isNullOrEmpty(job.jobAuthorization().encryptedInitialExportAuthData()));
```

Comment:
nit: break up this statement to fit within col limits

Reply:
Done